### DISCUSSION TITLE
Why does the openssl ccm module not support multiple updates, segmented encryption and decryption

### DISCUSSION THREAD

**Discussion by beebeeyoung** [openssl/openssl#25719] [Q&A] [answered]

**Comment by beebeeyoung** [comment]
int ccm_encrypt(unsigned char *plaintext, int plaintext_len,
                unsigned char *aad, int aad_len,
                unsigned char *key,
                unsigned char *iv,
                unsigned char *ciphertext,
                unsigned char *tag)
{
    EVP_CIPHER_CTX *ctx;

    int len;

    int ciphertext_len;


    /* Create and initialise the context */
    if(!(ctx = EVP_CIPHER_CTX_new()))
        handleErrors();

    /* Initialise the encryption operation. */
    if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_ccm(), NULL, NULL, NULL))
        handleErrors();

    /*
     * Setting IV len to 7. Not strictly necessary as this is the default
     * but shown here for the purposes of this example.
     */
    if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_CCM_SET_IVLEN, 7, NULL))
        handleErrors();

    /* Set tag length */
    EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_CCM_SET_TAG, 14, NULL);

    /* Initialise key and IV */
    if(1 != EVP_EncryptInit_ex(ctx, NULL, NULL, key, iv))
        handleErrors();

    /* Provide the total plaintext length */
    if(1 != EVP_EncryptUpdate(ctx, NULL, &len, NULL, plaintext_len))
        handleErrors();

    /* Provide any AAD data. This can be called zero or one times as required */
    if(1 != EVP_EncryptUpdate(ctx, NULL, &len, aad, aad_len))
        handleErrors();

    /*
     * Provide the message to be encrypted, and obtain the encrypted output.
     * EVP_EncryptUpdate can only be called once for this.
     */
    if(1 != EVP_EncryptUpdate(ctx, ciphertext, &len, plaintext, plaintext_len))
        handleErrors();
    ciphertext_len = len;

    /*
     * Finalise the encryption. Normally ciphertext bytes may be written at
     * this stage, but this does not occur in CCM mode.
     */
    if(1 != EVP_EncryptFinal_ex(ctx, ciphertext + len, &len))
        handleErrors();
    ciphertext_len += len;

    /* Get the tag */
    if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_CCM_GET_TAG, 14, tag))
        handleErrors();

    /* Clean up */
    EVP_CIPHER_CTX_free(ctx);

    return ciphertext_len;
}

**Reply by beebeeyoung** [reply]
This is inconvenient; For example, with only 1GB of available memory, a 2GB file cannot be encrypted using EVP_EncryptUpdate in one call; Only multiple calls to EVP_EncryptUpdate can complete the encryption operation, but for CCM encryption mode, multiple calls to EVP_EncryptUpdate to encrypt data are not supported.

**Reply by t8m** [reply, marked as answer]
The CCM mode requires knowing the length of the message before the encryption. That makes the restriction for only one EVP_EncryptUpdate() call obvious. It is also not a good idea to encrypt and authenticate large chunks of data in one go as it requires buffering all the data in memory. Usually a file encryption formats split large files into multiple segments which are encrypted separately with each segment having a separate tag and computing an agregate MAC from each tag to preserve the integrity of the whole file. It would be possible, for provider based implementations, to add EVP_Encrypt/DecryptUpdate_ex() functions that would take size_t arguments for the lengths which would allow for encrypting larger data in one go however that would still require loading all the data into the memory (or doing mmap) and the system would have to be 64bit one.
